{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Akintayo\\AppData\\Local\\Temp\\ipykernel_17260\\1010761004.py:1: DtypeWarning: Columns (0,1,2,3,6,10,14,26,29,30,31,37,38,39,40,41,44,45,46,47,48,49,50,55,56,57,58,59,61,63,66,67,68,72,75,79,80,81,82,83,84,87,88,89,90,91,92,98,99,100) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  Metagenomic= pd.read_csv(r'C:\\Users\\Akintayo\\Desktop\\Ulcer\\General Folder\\Metagenomic_data.csv')\n"
     ]
    }
   ],
   "source": [
    "Metagenomic= pd.read_csv(r'C:\\Users\\Akintayo\\Desktop\\Ulcer\\General Folder\\Metagenomic_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Superclass</th>\n",
       "      <th>Class</th>\n",
       "      <th>Subclass</th>\n",
       "      <th>Subsystem Name</th>\n",
       "      <th>Gene Count</th>\n",
       "      <th>Role Count</th>\n",
       "      <th>Active</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>...</th>\n",
       "      <th>E-value.1</th>\n",
       "      <th>Pathway ID.1</th>\n",
       "      <th>Pathway Name.1</th>\n",
       "      <th>Pathway Class.1</th>\n",
       "      <th>Annotation.3</th>\n",
       "      <th>Unique Genome Count.1</th>\n",
       "      <th>Unique Gene Count.1</th>\n",
       "      <th>Unique EC Count.1</th>\n",
       "      <th>EC Conservation.1</th>\n",
       "      <th>Gene Conservation.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>METABOLISM</td>\n",
       "      <td>Nucleosides and Nucleotides</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adenosyl nucleosidases</td>\n",
       "      <td>29.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>active</td>\n",
       "      <td>Bacteria SRX12677509</td>\n",
       "      <td>2.45918</td>\n",
       "      <td>2.45918.con.0001</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENERGY</td>\n",
       "      <td>Energy and Precursor Metabolites Generation</td>\n",
       "      <td>Central Metabolism</td>\n",
       "      <td>Pyruvate metabolism I: anaplerotic reactions, PEP</td>\n",
       "      <td>53.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>active</td>\n",
       "      <td>Bacteria SRX12677509</td>\n",
       "      <td>2.45918</td>\n",
       "      <td>2.45918.con.0001</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>METABOLISM</td>\n",
       "      <td>Fatty Acids, Lipids, and Isoprenoids</td>\n",
       "      <td>Fatty acids</td>\n",
       "      <td>Fatty acid synthesis</td>\n",
       "      <td>78.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>active</td>\n",
       "      <td>Bacteria SRX12677509</td>\n",
       "      <td>2.45918</td>\n",
       "      <td>2.45918.con.0001</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MEMBRANE TRANSPORT</td>\n",
       "      <td>Membrane Transport</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Utilization systems for glycans and polysaccha...</td>\n",
       "      <td>676.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>active</td>\n",
       "      <td>Bacteria SRX12677509</td>\n",
       "      <td>2.45918</td>\n",
       "      <td>2.45918.con.0001</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>METABOLISM</td>\n",
       "      <td>Fatty Acids, Lipids, and Isoprenoids</td>\n",
       "      <td>Fatty acids</td>\n",
       "      <td>Fatty Acid Biosynthesis cluster</td>\n",
       "      <td>46.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>active</td>\n",
       "      <td>Bacteria SRX12677509</td>\n",
       "      <td>2.45918</td>\n",
       "      <td>2.45918.con.0001</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593122</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria SRX12677528shotgun metagenomics</td>\n",
       "      <td>2.41875</td>\n",
       "      <td>2.41875.con.1263</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593123</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria SRX12677528shotgun metagenomics</td>\n",
       "      <td>2.41875</td>\n",
       "      <td>2.41875.con.1264</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593124</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria SRX12677528shotgun metagenomics</td>\n",
       "      <td>2.41875</td>\n",
       "      <td>2.41875.con.1265</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593125</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria SRX12677528shotgun metagenomics</td>\n",
       "      <td>2.41875</td>\n",
       "      <td>2.41875.con.1266</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593126</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria SRX12677528shotgun metagenomics</td>\n",
       "      <td>2.41875</td>\n",
       "      <td>2.41875.con.1267</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>593127 rows Ã— 106 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Superclass                                        Class  \\\n",
       "0               METABOLISM                  Nucleosides and Nucleotides   \n",
       "1                   ENERGY  Energy and Precursor Metabolites Generation   \n",
       "2               METABOLISM         Fatty Acids, Lipids, and Isoprenoids   \n",
       "3       MEMBRANE TRANSPORT                           Membrane Transport   \n",
       "4               METABOLISM         Fatty Acids, Lipids, and Isoprenoids   \n",
       "...                    ...                                          ...   \n",
       "593122                 NaN                                          NaN   \n",
       "593123                 NaN                                          NaN   \n",
       "593124                 NaN                                          NaN   \n",
       "593125                 NaN                                          NaN   \n",
       "593126                 NaN                                          NaN   \n",
       "\n",
       "                  Subclass                                     Subsystem Name  \\\n",
       "0                      NaN                             Adenosyl nucleosidases   \n",
       "1       Central Metabolism  Pyruvate metabolism I: anaplerotic reactions, PEP   \n",
       "2              Fatty acids                               Fatty acid synthesis   \n",
       "3                      NaN  Utilization systems for glycans and polysaccha...   \n",
       "4              Fatty acids                    Fatty Acid Biosynthesis cluster   \n",
       "...                    ...                                                ...   \n",
       "593122                 NaN                                                NaN   \n",
       "593123                 NaN                                                NaN   \n",
       "593124                 NaN                                                NaN   \n",
       "593125                 NaN                                                NaN   \n",
       "593126                 NaN                                                NaN   \n",
       "\n",
       "        Gene Count  Role Count  Active  \\\n",
       "0             29.0         3.0  active   \n",
       "1             53.0         7.0  active   \n",
       "2             78.0        10.0  active   \n",
       "3            676.0         7.0  active   \n",
       "4             46.0         5.0  active   \n",
       "...            ...         ...     ...   \n",
       "593122         NaN         NaN     NaN   \n",
       "593123         NaN         NaN     NaN   \n",
       "593124         NaN         NaN     NaN   \n",
       "593125         NaN         NaN     NaN   \n",
       "593126         NaN         NaN     NaN   \n",
       "\n",
       "                                          Genome  Genome ID         Accession  \\\n",
       "0                           Bacteria SRX12677509    2.45918  2.45918.con.0001   \n",
       "1                           Bacteria SRX12677509    2.45918  2.45918.con.0001   \n",
       "2                           Bacteria SRX12677509    2.45918  2.45918.con.0001   \n",
       "3                           Bacteria SRX12677509    2.45918  2.45918.con.0001   \n",
       "4                           Bacteria SRX12677509    2.45918  2.45918.con.0001   \n",
       "...                                          ...        ...               ...   \n",
       "593122  Bacteria SRX12677528shotgun metagenomics    2.41875  2.41875.con.1263   \n",
       "593123  Bacteria SRX12677528shotgun metagenomics    2.41875  2.41875.con.1264   \n",
       "593124  Bacteria SRX12677528shotgun metagenomics    2.41875  2.41875.con.1265   \n",
       "593125  Bacteria SRX12677528shotgun metagenomics    2.41875  2.41875.con.1266   \n",
       "593126  Bacteria SRX12677528shotgun metagenomics    2.41875  2.41875.con.1267   \n",
       "\n",
       "        ... E-value.1  Pathway ID.1  Pathway Name.1 Pathway Class.1  \\\n",
       "0       ...       NaN           NaN             NaN             NaN   \n",
       "1       ...       NaN           NaN             NaN             NaN   \n",
       "2       ...       NaN           NaN             NaN             NaN   \n",
       "3       ...       NaN           NaN             NaN             NaN   \n",
       "4       ...       NaN           NaN             NaN             NaN   \n",
       "...     ...       ...           ...             ...             ...   \n",
       "593122  ...       NaN           NaN             NaN             NaN   \n",
       "593123  ...       NaN           NaN             NaN             NaN   \n",
       "593124  ...       NaN           NaN             NaN             NaN   \n",
       "593125  ...       NaN           NaN             NaN             NaN   \n",
       "593126  ...       NaN           NaN             NaN             NaN   \n",
       "\n",
       "       Annotation.3 Unique Genome Count.1  Unique Gene Count.1  \\\n",
       "0               NaN                   NaN                  NaN   \n",
       "1               NaN                   NaN                  NaN   \n",
       "2               NaN                   NaN                  NaN   \n",
       "3               NaN                   NaN                  NaN   \n",
       "4               NaN                   NaN                  NaN   \n",
       "...             ...                   ...                  ...   \n",
       "593122          NaN                   NaN                  NaN   \n",
       "593123          NaN                   NaN                  NaN   \n",
       "593124          NaN                   NaN                  NaN   \n",
       "593125          NaN                   NaN                  NaN   \n",
       "593126          NaN                   NaN                  NaN   \n",
       "\n",
       "        Unique EC Count.1  EC Conservation.1 Gene Conservation.1  \n",
       "0                     NaN                NaN                 NaN  \n",
       "1                     NaN                NaN                 NaN  \n",
       "2                     NaN                NaN                 NaN  \n",
       "3                     NaN                NaN                 NaN  \n",
       "4                     NaN                NaN                 NaN  \n",
       "...                   ...                ...                 ...  \n",
       "593122                NaN                NaN                 NaN  \n",
       "593123                NaN                NaN                 NaN  \n",
       "593124                NaN                NaN                 NaN  \n",
       "593125                NaN                NaN                 NaN  \n",
       "593126                NaN                NaN                 NaN  \n",
       "\n",
       "[593127 rows x 106 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Metagenomic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 593127 entries, 0 to 593126\n",
      "Columns: 106 entries, Superclass to Gene Conservation.1\n",
      "dtypes: float64(49), object(57)\n",
      "memory usage: 479.7+ MB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(None,\n",
       "            Superclass                                        Class  \\\n",
       " 0          METABOLISM                  Nucleosides and Nucleotides   \n",
       " 1              ENERGY  Energy and Precursor Metabolites Generation   \n",
       " 2          METABOLISM         Fatty Acids, Lipids, and Isoprenoids   \n",
       " 3  MEMBRANE TRANSPORT                           Membrane Transport   \n",
       " 4          METABOLISM         Fatty Acids, Lipids, and Isoprenoids   \n",
       " \n",
       "              Subclass                                     Subsystem Name  \\\n",
       " 0                 NaN                             Adenosyl nucleosidases   \n",
       " 1  Central Metabolism  Pyruvate metabolism I: anaplerotic reactions, PEP   \n",
       " 2         Fatty acids                               Fatty acid synthesis   \n",
       " 3                 NaN  Utilization systems for glycans and polysaccha...   \n",
       " 4         Fatty acids                    Fatty Acid Biosynthesis cluster   \n",
       " \n",
       "    Gene Count  Role Count  Active                Genome  Genome ID  \\\n",
       " 0        29.0         3.0  active  Bacteria SRX12677509    2.45918   \n",
       " 1        53.0         7.0  active  Bacteria SRX12677509    2.45918   \n",
       " 2        78.0        10.0  active  Bacteria SRX12677509    2.45918   \n",
       " 3       676.0         7.0  active  Bacteria SRX12677509    2.45918   \n",
       " 4        46.0         5.0  active  Bacteria SRX12677509    2.45918   \n",
       " \n",
       "           Accession  ... E-value.1  Pathway ID.1  Pathway Name.1  \\\n",
       " 0  2.45918.con.0001  ...       NaN           NaN             NaN   \n",
       " 1  2.45918.con.0001  ...       NaN           NaN             NaN   \n",
       " 2  2.45918.con.0001  ...       NaN           NaN             NaN   \n",
       " 3  2.45918.con.0001  ...       NaN           NaN             NaN   \n",
       " 4  2.45918.con.0001  ...       NaN           NaN             NaN   \n",
       " \n",
       "   Pathway Class.1 Annotation.3 Unique Genome Count.1  Unique Gene Count.1  \\\n",
       " 0             NaN          NaN                   NaN                  NaN   \n",
       " 1             NaN          NaN                   NaN                  NaN   \n",
       " 2             NaN          NaN                   NaN                  NaN   \n",
       " 3             NaN          NaN                   NaN                  NaN   \n",
       " 4             NaN          NaN                   NaN                  NaN   \n",
       " \n",
       "    Unique EC Count.1  EC Conservation.1 Gene Conservation.1  \n",
       " 0                NaN                NaN                 NaN  \n",
       " 1                NaN                NaN                 NaN  \n",
       " 2                NaN                NaN                 NaN  \n",
       " 3                NaN                NaN                 NaN  \n",
       " 4                NaN                NaN                 NaN  \n",
       " \n",
       " [5 rows x 106 columns])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display basic information about the dataset and first few rows for initial inspection\n",
    "data_info = Metagenomic.info()\n",
    "data_head = Metagenomic.head()\n",
    "data_info, Metagenomic.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(Class\n",
       " Protein Synthesis                                               3332\n",
       " Amino Acids and Derivatives                                     2469\n",
       " Cofactors, Vitamins, Prosthetic Groups                          2442\n",
       " Energy and Precursor Metabolites Generation                     1834\n",
       " Membrane Transport                                              1777\n",
       " Stress Response, Defense and Virulence                          1760\n",
       " DNA Processing                                                  1733\n",
       " Carbohydrates                                                   1526\n",
       " Cell Cycle, Cell Division and Death                             1333\n",
       " RNA Processing                                                  1240\n",
       " Fatty Acids, Lipids, and Isoprenoids                            1035\n",
       " Nucleosides and Nucleotides                                      912\n",
       " Protein Fate (folding, modification, targeting, degradation)     796\n",
       " Respiration                                                      760\n",
       " Cell Envelope, Capsule and Slime layer                           536\n",
       " Prophages, Transposable elements, Plasmids                       429\n",
       " Metabolite damage and its repair or mitigation                   293\n",
       " Prokaryotic cell type differentiation                            291\n",
       " Phosphate Metabolism                                             213\n",
       " Clustering-based subsystems                                      179\n",
       " Regulation and Cell signaling                                    160\n",
       " Secondary Metabolism                                             133\n",
       " Miscellaneous                                                    101\n",
       " Iron acquisition and metabolism                                   94\n",
       " Experimental Subsystems                                           44\n",
       " Motility and Chemotaxis                                           37\n",
       " Sulfur Metabolism                                                  8\n",
       " Microbial communities                                              5\n",
       " Nitrogen Metabolism                                                2\n",
       " Name: count, dtype: int64,\n",
       " Superclass\n",
       " METABOLISM                             9037\n",
       " PROTEIN PROCESSING                     4128\n",
       " ENERGY                                 2594\n",
       " CELLULAR PROCESSES                     1845\n",
       " MEMBRANE TRANSPORT                     1777\n",
       " STRESS RESPONSE, DEFENSE, VIRULENCE    1760\n",
       " DNA PROCESSING                         1733\n",
       " RNA PROCESSING                         1240\n",
       " CELL ENVELOPE                           536\n",
       " MISCELLANEOUS                           530\n",
       " REGULATION AND CELL SIGNALING           160\n",
       " Name: count, dtype: int64)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summary of unique classes and counts within the dataset to understand predominant functional categories\n",
    "class_counts = Metagenomic['Class'].value_counts()\n",
    "\n",
    "# Examine the distribution of Superclasses as well\n",
    "superclass_counts = Metagenomic['Superclass'].value_counts()\n",
    "\n",
    "class_counts, superclass_counts\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identify specific bacteria and genes linked to gut ulcers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([nan, 'tmRNA-binding protein SmpB',\n",
       "        'Translation elongation factor G',\n",
       "        'Translation elongation factor LepA',\n",
       "        'Trk potassium uptake system protein TrkH',\n",
       "        '1-deoxy-D-xylulose 5-phosphate reductoisomerase (EC 1.1.1.267)',\n",
       "        'SSU ribosomal protein S10p (S20e)',\n",
       "        'Oxygen-independent coproporphyrinogen-III oxidase-like protein YggW',\n",
       "        'SSU ribosomal protein S1p', 'Putative pre-16S rRNA nuclease YqgF',\n",
       "        'Glycine betaine ABC transport system, ATP-binding protein OpuAA (EC 3.6.3.32)',\n",
       "        'Aerotolerance protein BatE',\n",
       "        'Nucleoside 5-triphosphatase RdgB (dHAPTP, dITP, XTP-specific) (EC 3.6.1.66)',\n",
       "        'DNA topoisomerase IV subunit A (EC 5.99.1.3)',\n",
       "        'UDP-N-acetylglucosamine 1-carboxyvinyltransferase (EC 2.5.1.7)',\n",
       "        'SSU ribosomal protein S12p (S23e)',\n",
       "        'LSU ribosomal protein L6p (L9e)',\n",
       "        'Lactoylglutathione lyase (EC 4.4.1.5)',\n",
       "        'DNA topoisomerase IV subunit B (EC 5.99.1.3)',\n",
       "        'Isoleucyl-tRNA synthetase (EC 6.1.1.5)', 'SSU rRNA',\n",
       "        'DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)',\n",
       "        '16S rRNA (uracil(1498)-N(3))-methyltransferase (EC 2.1.1.193)',\n",
       "        'DNA gyrase subunit A (EC 5.99.1.3)',\n",
       "        'Cytoplasmic copper homeostasis protein cutC',\n",
       "        'Heat shock protein GrpE', 'Chaperone protein DnaK',\n",
       "        'hypothetical protein PA3071',\n",
       "        'tRNA t(6)A37-methylthiotransferase (EC 2.8.4.5)',\n",
       "        \"DNA-directed RNA polymerase beta' subunit (EC 2.7.7.6)\",\n",
       "        'Transcription termination factor Rho', 'LSU rRNA',\n",
       "        \"16S rRNA (cytidine(1402)-2'-O)-methyltransferase (EC 2.1.1.198)\",\n",
       "        'Alanine racemase (EC 5.1.1.1)',\n",
       "        'CDP-diacylglycerol--glycerol-3-phosphate 3-phosphatidyltransferase (EC 2.7.8.5)',\n",
       "        'Cold shock protein of CSP family',\n",
       "        'Translation elongation factor Tu',\n",
       "        'D-alanine--D-alanine ligase (EC 6.3.2.4)',\n",
       "        'Aerotolerance protein BatC', 'Aerotolerance protein BatB',\n",
       "        'DNA gyrase subunit B (EC 5.99.1.3)',\n",
       "        'Ribosome LSU-associated GTP-binding protein HflX',\n",
       "        'Cardiolipin synthase, bacterial type ClsA', 'Aquaporin Z',\n",
       "        'Hemerythrin domain protein',\n",
       "        'Macrolide resistance, MFS efflux pump => Mef(A)',\n",
       "        'Aerotolerance protein BatD', 'Oleate hydratase (EC 4.2.1.53)',\n",
       "        'Glycerophosphoryl diester phosphodiesterase (EC 3.1.4.46)',\n",
       "        'Chaperone protein DnaJ',\n",
       "        'Trk potassium uptake system protein TrkA',\n",
       "        'Ribosomal protein L11 methyltransferase',\n",
       "        'Uncharacterized membrane protein, YraQ family',\n",
       "        'Choloylglycine hydrolase (EC 3.5.1.24)',\n",
       "        'High-affinity choline uptake protein BetT',\n",
       "        'Glycine betaine ABC transport system, permease protein OpuAB',\n",
       "        'MarC family integral membrane protein',\n",
       "        'Redox-active disulfide protein 2',\n",
       "        'Arsenical resistance operon repressor',\n",
       "        'Copper homeostasis protein CutF precursor',\n",
       "        'Dihydropteroate synthase (EC 2.5.1.15)',\n",
       "        'Dihydrofolate reductase (EC 1.5.1.3)',\n",
       "        '3-oxoacyl-[acyl-carrier-protein] synthase, KASII (EC 2.3.1.179)',\n",
       "        'Enoyl-[acyl-carrier-protein] reductase [FMN, NADH] (EC 1.3.1.9), FabK => refractory to triclosan',\n",
       "        'Superoxide dismutase [Fe] (EC 1.15.1.1)',\n",
       "        'FIG00936810: hypothetical protein',\n",
       "        'Repair of Iron Centers di-iron protein',\n",
       "        'weak similarity to aminoglycoside N(3)-acetyltransferase',\n",
       "        'Heat-inducible transcription repressor HrcA',\n",
       "        'ABC-F type ribosomal protection protein => Msr(D)',\n",
       "        '23S rRNA (adenine(2058)-N(6))-dimethyltransferase (EC 2.1.1.184) => Erm(G)',\n",
       "        'Aminoglycoside 9-nucleotidyltransferase (EC 2.7.7.-) => ANT(9)-I',\n",
       "        'Tetracycline resistance, ribosomal protection type => Tet(O)',\n",
       "        'Enoyl-[acyl-carrier-protein] reductase [NADH] (EC 1.3.1.9)',\n",
       "        'MoxR-like ATPase in aerotolerance operon',\n",
       "        'Arsenate reductase (EC 1.20.4.4) thioredoxin-coupled, LMWP family',\n",
       "        'Cytochrome c551 peroxidase (EC 1.11.1.5)',\n",
       "        'Glycine betaine transporter OpuD', 'Aerotolerance protein BatA',\n",
       "        'Catalase KatE (EC 1.11.1.6)', 'UPF0301 protein YqgE',\n",
       "        'Aminoglycoside 6-nucleotidyltransferase (EC 2.7.7.-) => ANT(6)-I',\n",
       "        'Outer membrane protein A precursor',\n",
       "        'Arsenical-resistance protein ACR3',\n",
       "        'Sphingomyelinase (EC 3.1.4.12)',\n",
       "        'Class A beta-lactamase (EC 3.5.2.6)',\n",
       "        'Glutathione peroxidase (EC 1.11.1.9)', 'RNA-binding protein Hfq',\n",
       "        'Universal stress protein family',\n",
       "        'Putative FabK-like enoyl-[acyl-carrier-protein] reductase',\n",
       "        'Glycine betaine ABC transport system, glycine betaine-binding protein OpuAC',\n",
       "        'Ribonuclease PH (EC 2.7.7.56)', 'Probable metallo-hydrolase YqgX',\n",
       "        'Glutathione biosynthesis bifunctional protein gshF (EC 6.3.2.2)(EC 6.3.2.3)',\n",
       "        'Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)',\n",
       "        'Tetracycline resistance, ribosomal protection type => Tet(Q)',\n",
       "        'UPF0056 inner membrane protein MarC'], dtype=object),\n",
       " array(['3-oxoacyl-[acyl-carrier-protein] synthase, KASII (EC 2.3.1.179)',\n",
       "        'Acyl carrier protein', 'hypothetical protein',\n",
       "        'Putative ion-channel protein',\n",
       "        'Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family',\n",
       "        'Polysaccharide deacetylase',\n",
       "        'Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family',\n",
       "        'Glutamate racemase (EC 5.1.1.3)',\n",
       "        'Putative TonB-dependent receptor', 'Putative anti-sigma factor',\n",
       "        \"5'-methylthioadenosine nucleosidase (EC 3.2.2.16) @ S-adenosylhomocysteine nucleosidase (EC 3.2.2.9)\",\n",
       "        'Immunoreactive 53 kDa antigen PG123',\n",
       "        '2-amino-3-ketobutyrate coenzyme A ligase (EC 2.3.1.29)',\n",
       "        'L-threonine 3-dehydrogenase (EC 1.1.1.103)',\n",
       "        'AP endonuclease domain protein', 'Rhomboid family protein',\n",
       "        'beta-galactosidase (EC 3.2.1.23)',\n",
       "        'tRNA-specific 2-thiouridylase MnmA (EC 2.8.1.13)',\n",
       "        'Undecaprenyl diphosphate synthase, for quinones (EC 2.5.1.31) @ Decaprenyl diphosphate synthase (EC 2.5.1.91)',\n",
       "        'Deoxyribose-phosphate aldolase (EC 4.1.2.4)',\n",
       "        'D-aminoacyl-tRNA deacylase (EC 3.1.1.96)',\n",
       "        'Xylulose-5-phosphate phosphoketolase (EC 4.1.2.9) @ Fructose-6-phosphate phosphoketolase (EC 4.1.2.22)',\n",
       "        'UDP-N-acetylmuramoyl-tripeptide--D-alanyl-D-alanine ligase (EC 6.3.2.10)',\n",
       "        'Substrate-specific component RibU of riboflavin ECF transporter',\n",
       "        'UDP-N-acetylmuramate--L-alanine ligase (EC 6.3.2.8)',\n",
       "        'DNA-binding protein SpoVG',\n",
       "        'N-acetylglucosamine-1-phosphate uridyltransferase (EC 2.7.7.23)',\n",
       "        'Cell division protein DivIC (FtsB), stabilizes FtsL against RasP cleavage',\n",
       "        'Sulfate and thiosulfate import ATP-binding protein CysA (EC 3.6.3.25)',\n",
       "        'Tellurium resistance protein TerD',\n",
       "        'Phosphoribosylamine--glycine ligase (EC 6.3.4.13)',\n",
       "        'Serine acetyltransferase (EC 2.3.1.30)',\n",
       "        'Phosphatidylglycerophosphatase A (EC 3.1.3.27)',\n",
       "        'Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator',\n",
       "        '6,7-dimethyl-8-ribityllumazine synthase (EC 2.5.1.78)',\n",
       "        'FAD:protein FMN transferase (EC 2.7.1.180)',\n",
       "        'N-acetylgalactosamine-6-sulfatase',\n",
       "        'Two-component system sensor histidine kinase',\n",
       "        'Uncharacterized SAM-dependent O-methyltransferase',\n",
       "        'Beta-xylosidase', 'RhuM',\n",
       "        'Protein translocase subunit SecD / Protein translocase subunit SecF',\n",
       "        'S-ribosylhomocysteine lyase (EC 4.4.1.21) @ Autoinducer-2 production protein LuxS',\n",
       "        'Dihydrolipoamide dehydrogenase (EC 1.8.1.4)',\n",
       "        'Putative large secreted protein SCO0341',\n",
       "        'Putative glycosyl hydrolase of unknown function (DUF1680)',\n",
       "        'Secreted protein', 'glycoside hydrolase, family 28',\n",
       "        'Alpha-1,2-mannosidase', 'tmRNA-binding protein SmpB',\n",
       "        'Translation elongation factor G',\n",
       "        'Translation elongation factor LepA',\n",
       "        'Trk potassium uptake system protein TrkH',\n",
       "        '1-deoxy-D-xylulose 5-phosphate reductoisomerase (EC 1.1.1.267)',\n",
       "        'SSU ribosomal protein S10p (S20e)',\n",
       "        'Oxygen-independent coproporphyrinogen-III oxidase-like protein YggW',\n",
       "        'SSU ribosomal protein S1p', 'Putative pre-16S rRNA nuclease YqgF',\n",
       "        'Glycine betaine ABC transport system, ATP-binding protein OpuAA (EC 3.6.3.32)',\n",
       "        'Aerotolerance protein BatE',\n",
       "        'Nucleoside 5-triphosphatase RdgB (dHAPTP, dITP, XTP-specific) (EC 3.6.1.66)',\n",
       "        'DNA topoisomerase IV subunit A (EC 5.99.1.3)',\n",
       "        'UDP-N-acetylglucosamine 1-carboxyvinyltransferase (EC 2.5.1.7)',\n",
       "        'SSU ribosomal protein S12p (S23e)',\n",
       "        'LSU ribosomal protein L6p (L9e)',\n",
       "        'Lactoylglutathione lyase (EC 4.4.1.5)',\n",
       "        'DNA topoisomerase IV subunit B (EC 5.99.1.3)',\n",
       "        'Isoleucyl-tRNA synthetase (EC 6.1.1.5)',\n",
       "        'SSU rRNA ## 16S rRNA, small subunit ribosomal RNA',\n",
       "        'DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)',\n",
       "        '16S rRNA (uracil(1498)-N(3))-methyltransferase (EC 2.1.1.193)',\n",
       "        'DNA gyrase subunit A (EC 5.99.1.3)',\n",
       "        'Cytoplasmic copper homeostasis protein CutC',\n",
       "        'Heat shock protein GrpE', 'Chaperone protein DnaK',\n",
       "        'hypothetical protein PA3071',\n",
       "        'tRNA t(6)A37-methylthiotransferase (EC 2.8.4.5)',\n",
       "        \"DNA-directed RNA polymerase beta' subunit (EC 2.7.7.6)\",\n",
       "        'Transcription termination factor Rho',\n",
       "        'LSU rRNA # 23S rRNA, large subunit ribosomal RNA - 5 prime truncation',\n",
       "        \"16S rRNA (cytidine(1402)-2'-O)-methyltransferase (EC 2.1.1.198)\",\n",
       "        'SSU rRNA # 16S rRNA, small subunit ribosomal RNA - 3 prime truncation',\n",
       "        'UDP-N-acetylmuramoyl-tripeptide--D-alanyl-D-alanine ligase (EC 6.3.2.10) / Alanine racemase (EC 5.1.1.1)',\n",
       "        'CDP-diacylglycerol--glycerol-3-phosphate 3-phosphatidyltransferase (EC 2.7.8.5)',\n",
       "        'hypothetical domain / Cold shock protein of CSP family',\n",
       "        'Translation elongation factor Tu',\n",
       "        'D-alanine--D-alanine ligase (EC 6.3.2.4)',\n",
       "        'Aerotolerance protein BatC', 'Aerotolerance protein BatB',\n",
       "        'DNA gyrase subunit B (EC 5.99.1.3)',\n",
       "        'Ribosome LSU-associated GTP-binding protein HflX',\n",
       "        'SSU rRNA # 16S rRNA, small subunit ribosomal RNA - 5 prime truncation',\n",
       "        'Cardiolipin synthase, bacterial type ClsA', 'Aquaporin Z',\n",
       "        'Hemerythrin domain protein',\n",
       "        'Macrolide resistance, MFS efflux pump => Mef(A)',\n",
       "        'Aerotolerance protein BatD', 'Oleate hydratase (EC 4.2.1.53)',\n",
       "        'Glycerophosphoryl diester phosphodiesterase (EC 3.1.4.46)',\n",
       "        'Chaperone protein DnaJ',\n",
       "        'Trk potassium uptake system protein TrkA',\n",
       "        'Ribosomal protein L11 methyltransferase',\n",
       "        'Alanine racemase (EC 5.1.1.1)',\n",
       "        'Uncharacterized membrane protein, YraQ family',\n",
       "        'Choloylglycine hydrolase (EC 3.5.1.24)',\n",
       "        'High-affinity choline uptake protein BetT',\n",
       "        'LSU rRNA # 23S rRNA, large subunit ribosomal RNA - 5 prime truncation - 3 prime truncation',\n",
       "        'Glycine betaine ABC transport system, permease protein OpuAB',\n",
       "        'MarC family integral membrane protein',\n",
       "        'Redox-active disulfide protein 2',\n",
       "        'Arsenical resistance operon repressor',\n",
       "        'Copper homeostasis protein CutF precursor / Lipoprotein NlpE involeved in surface adhesion',\n",
       "        'Dihydropteroate synthase (EC 2.5.1.15)',\n",
       "        'LSU rRNA ## 23S rRNA, large subunit ribosomal RNA',\n",
       "        'Dihydrofolate reductase (EC 1.5.1.3)',\n",
       "        'Enoyl-[acyl-carrier-protein] reductase [FMN, NADH] (EC 1.3.1.9), FabK => refractory to triclosan',\n",
       "        'Superoxide dismutase [Fe] (EC 1.15.1.1) / Rubredoxin',\n",
       "        'FIG00936810: hypothetical protein',\n",
       "        'Repair of Iron Centers di-iron protein',\n",
       "        'weak similarity to aminoglycoside N(3)-acetyltransferase',\n",
       "        'Heat-inducible transcription repressor HrcA',\n",
       "        'LSU rRNA # 23S rRNA, large subunit ribosomal RNA - 3 prime truncation',\n",
       "        'ABC-F type ribosomal protection protein => Msr(D)',\n",
       "        '23S rRNA (adenine(2058)-N(6))-dimethyltransferase (EC 2.1.1.184) => Erm(G)',\n",
       "        'Aminoglycoside 9-nucleotidyltransferase (EC 2.7.7.-) => ANT(9)-I',\n",
       "        'Aerotolerance protein BatB / Aerotolerance protein BatC',\n",
       "        'Tetracycline resistance, ribosomal protection type => Tet(O)',\n",
       "        'Enoyl-[acyl-carrier-protein] reductase [NADH] (EC 1.3.1.9)',\n",
       "        'MoxR-like ATPase in aerotolerance operon',\n",
       "        'Arsenate reductase (EC 1.20.4.4) thioredoxin-coupled, LMWP family',\n",
       "        'Cytochrome c551 peroxidase (EC 1.11.1.5)',\n",
       "        'SSU rRNA # 16S rRNA, small subunit ribosomal RNA - 5 prime truncation - 3 prime truncation',\n",
       "        'Glycine betaine transporter OpuD', 'Aerotolerance protein BatA',\n",
       "        'Catalase KatE (EC 1.11.1.6)', 'UPF0301 protein YqgE',\n",
       "        'Aminoglycoside 6-nucleotidyltransferase (EC 2.7.7.-) => ANT(6)-I',\n",
       "        'Outer membrane protein A precursor',\n",
       "        'Arsenical-resistance protein ACR3',\n",
       "        'Sphingomyelinase (EC 3.1.4.12)',\n",
       "        'Class A beta-lactamase (EC 3.5.2.6)',\n",
       "        'Glutathione peroxidase (EC 1.11.1.9) @ Thioredoxin peroxidase (EC 1.11.1.15)',\n",
       "        'RNA-binding protein Hfq', 'Universal stress protein family',\n",
       "        'Putative FabK-like enoyl-[acyl-carrier-protein] reductase',\n",
       "        'Glycine betaine ABC transport system, glycine betaine-binding protein OpuAC',\n",
       "        'Choloylglycine hydrolase', 'Ribonuclease PH (EC 2.7.7.56)',\n",
       "        'Probable metallo-hydrolase YqgX',\n",
       "        'Cold shock protein of CSP family',\n",
       "        'PaaD-like protein (DUF59) involved in Fe-S cluster assembly / Hemerythrin domain protein',\n",
       "        'Glutathione biosynthesis bifunctional protein gshF (EC 6.3.2.2)(EC 6.3.2.3)',\n",
       "        'Sialic acid-specific 9-O-acetylesterase', 'Serine protease',\n",
       "        'Biotin synthase (EC 2.8.1.6)',\n",
       "        'Adenosylmethionine-8-amino-7-oxononanoate aminotransferase (EC 2.6.1.62)',\n",
       "        'ABC-type antimicrobial peptide transport system, permease component BT1215',\n",
       "        'Ferrous iron transporter-associated protein FeoA / Ferrous iron transporter FeoB',\n",
       "        'Pantoate--beta-alanine ligase (EC 6.3.2.1)',\n",
       "        'Dipeptidyl carboxypeptidase Dcp (EC 3.4.15.5)',\n",
       "        'Putative manganese efflux pump MntP',\n",
       "        'Predicted cobalt transporter in Bacteroides_Porphyromonas',\n",
       "        'no significant homology.',\n",
       "        'Methionine aminopeptidase (EC 3.4.11.18)',\n",
       "        'DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)',\n",
       "        'Nucleoside permease NupG',\n",
       "        'UDP-3-O-[3-hydroxymyristoyl] glucosamine N-acyltransferase (EC 2.3.1.191)',\n",
       "        'Aspartyl-tRNA synthetase (EC 6.1.1.12)',\n",
       "        'Mannose-1-phosphate guanylyltransferase (EC 2.7.7.13)',\n",
       "        '2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, beta subunit (EC 1.2.7.-)',\n",
       "        'Polysaccharide pyruvyl transferase',\n",
       "        'tRNA (guanine(37)-N(1))-methyltransferase (EC 2.1.1.228)',\n",
       "        'Pantothenate kinase type II, eukaryotic (EC 2.7.1.33)',\n",
       "        'Elongation factor G-like protein TM_1651',\n",
       "        'Two-component transcriptional response regulator RprY, OmpR family',\n",
       "        'Glycine dehydrogenase [decarboxylating] (glycine cleavage system P1 protein) (EC 1.4.4.2)',\n",
       "        'M. jannaschii predicted coding region MJ0455',\n",
       "        'D-3-phosphoglycerate dehydrogenase (EC 1.1.1.95)',\n",
       "        'Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)',\n",
       "        'Tryptophan synthase beta chain (EC 4.2.1.20)',\n",
       "        'tRNA-5-carboxymethylaminomethyl-2-thiouridine(34) synthesis protein MnmE',\n",
       "        'tRNA-(cytosine32)-2-thiocytidine synthetase TtcA',\n",
       "        'Serine hydroxymethyltransferase (EC 2.1.2.1)',\n",
       "        'Monofunctional biosynthetic peptidoglycan transglycosylase',\n",
       "        'Peptidase M64', 'putative ferric aerobactin receptor',\n",
       "        'Long-chain-fatty-acid--CoA ligase (EC 6.2.1.3)',\n",
       "        '2-C-methyl-D-erythritol 2,4-cyclodiphosphate synthase (EC 4.6.1.12)',\n",
       "        'Amidophosphoribosyltransferase (EC 2.4.2.14)',\n",
       "        'Lead, cadmium, zinc and mercury transporting ATPase (EC 3.6.3.3) (EC 3.6.3.5); Copper-translocating P-type ATPase (EC 3.6.3.4)',\n",
       "        'DinG family ATP-dependent helicase CPE1197',\n",
       "        'Methyl-accepting chemotaxis sensor/transducer protein',\n",
       "        'Peptide chain release factor 3',\n",
       "        'Histidyl-tRNA synthetase (EC 6.1.1.21)', 'Enolase (EC 4.2.1.11)',\n",
       "        'Transcriptional regulator, GntR family',\n",
       "        'Melibiose carrier protein, Na+/melibiose symporter',\n",
       "        'Seryl-tRNA synthetase (EC 6.1.1.11)',\n",
       "        'Chromosomal replication initiator protein DnaA',\n",
       "        'Uncharacterized S4 RNA-binding-domain protein YbcJ',\n",
       "        'Fumarate hydratase class I, beta region (EC 4.2.1.2); L(+)-tartrate dehydratase beta subunit (EC 4.2.1.32)',\n",
       "        'Argininosuccinate lyase (EC 4.3.2.1)',\n",
       "        'Fe-S oxidoreductase, related to NifB/MoaA family with PDZ N-terminal domain',\n",
       "        'Segregation and condensation protein A',\n",
       "        'Integrase/recombinase xerD', 'Cytidylate kinase (EC 2.7.4.25)',\n",
       "        'Septum site-determining protein MinD',\n",
       "        'Aluminum resistance protein',\n",
       "        'tRNA dimethylallyltransferase (EC 2.5.1.75)',\n",
       "        'DNA mismatch repair protein MutL',\n",
       "        'Domain similar to amino terminal domain of imidazole glycerol phosphate synthase cyclase subunit / Phosphoribosyl-AMP cyclohydrolase (EC 3.5.4.19) / Phosphoribosyl-ATP pyrophosphatase (EC 3.6.1.31)',\n",
       "        'Ferrous iron transporter FeoB',\n",
       "        'CDP-diacylglycerol--serine O-phosphatidyltransferase (EC 2.7.8.8)',\n",
       "        'Uncharacterized flavin-containing protein YtfP',\n",
       "        'Oxaloacetate decarboxylase beta chain (EC 4.1.1.3) @ Methylmalonyl-CoA decarboxylase, beta chain (EC 4.1.1.41)',\n",
       "        'Methenyltetrahydrofolate cyclohydrolase (EC 3.5.4.9) / Methylenetetrahydrofolate dehydrogenase (NADP+) (EC 1.5.1.5)',\n",
       "        'Undecaprenyl-diphosphatase (EC 3.6.1.27)',\n",
       "        'Polyribonucleotide nucleotidyltransferase (EC 2.7.7.8)',\n",
       "        'FIG069887: hypothetical protein',\n",
       "        'Formate efflux transporter FocA', 'UPF0348 protein family',\n",
       "        'Peptide chain release factor N(5)-glutamine methyltransferase (EC 2.1.1.297)',\n",
       "        '16S rRNA (guanine(966)-N(2))-methyltransferase (EC 2.1.1.171)',\n",
       "        'beta-1,4-glucanase (cellulase) (EC 3.2.1.4)',\n",
       "        'ATP-dependent DNA helicase RecG (EC 3.6.4.12)',\n",
       "        'Dihydroxyacetone kinase-like protein, phosphatase domain / Dihydroxyacetone kinase-like protein, kinase domain',\n",
       "        'Endonuclease III (EC 4.2.99.18)',\n",
       "        'Acetyltransferase, GNAT family',\n",
       "        '2-dehydropantoate 2-reductase (EC 1.1.1.169)',\n",
       "        'Electron transport complex protein RnfC',\n",
       "        'Electron transport complex protein RnfD', 'Rubrerythrin',\n",
       "        'Heterodimeric efflux ABC transporter, permease/ATP-binding subunit 1',\n",
       "        'Leader peptidase (Prepilin peptidase) (EC 3.4.23.43) / N-methyltransferase (EC 2.1.1.-)',\n",
       "        'alpha-galactosidase (EC 3.2.1.22)',\n",
       "        'Tyrosyl-tRNA synthetase (EC 6.1.1.1)',\n",
       "        'Aspartyl-tRNA synthetase (EC 6.1.1.12) @ Aspartyl-tRNA(Asn) synthetase (EC 6.1.1.23)',\n",
       "        '33 kDa chaperonin HslO', 'Glycosyl transferase, group 2 family',\n",
       "        'Alpha-D-GlcNAc alpha-1,2-L-rhamnosyltransferase (EC 2.4.1.-)',\n",
       "        'glycosyltransferase family 11',\n",
       "        'UDP-N-acetyl-D-glucosamine 6-dehydrogenase (EC 1.1.1.136)',\n",
       "        '4-keto-6-deoxy-N-Acetyl-D-hexosaminyl-(Lipid carrier) aminotransferase',\n",
       "        'tRNA/rRNA cytosine-C5-methyltransferase',\n",
       "        'Polyketide synthase modules and related proteins',\n",
       "        'Polymyxin resistance protein ArnC, glycosyl transferase (EC 2.4.-.-)',\n",
       "        'Bacillosamine/Legionaminic acid biosynthesis aminotransferase PglE; 4-keto-6-deoxy-N-Acetyl-D-hexosaminyl-(Lipid carrier) aminotransferase',\n",
       "        'Phosphonopyruvate decarboxylase (EC 4.1.1.82)',\n",
       "        'Cytidylyltransferase domain / Phosphoenolpyruvate phosphomutase (EC 5.4.2.9)',\n",
       "        '2-aminoethylphosphonate:pyruvate aminotransferase (EC 2.6.1.37)',\n",
       "        'Cell division protein FtsQ', 'Transposase',\n",
       "        'Xanthine phosphoribosyltransferase (EC 2.4.2.22)',\n",
       "        'Guanine deaminase (EC 3.5.4.3)', 'TonB-dependent receptor',\n",
       "        '2-dehydro-3-deoxygluconokinase (EC 2.7.1.45)',\n",
       "        'Glucan 1,4-alpha-glucosidase (EC 3.2.1.3)',\n",
       "        '4-alpha-glucanotransferase (amylomaltase) (EC 2.4.1.25)',\n",
       "        'Integrase', 'Phosphate transport system regulatory protein PhoU',\n",
       "        'tRNA pseudouridine(38-40) synthase (EC 5.4.99.12)',\n",
       "        'Heterodimeric efflux ABC transporter, permease/ATP-binding subunit 2',\n",
       "        'Flagellin protein FlaA', 'Phage integrase',\n",
       "        'Iron-sulfur cluster regulator IscR',\n",
       "        'Substrate-specific component FolT of folate ECF transporter',\n",
       "        'Biosynthetic Aromatic amino acid aminotransferase alpha (EC 2.6.1.57) @ Aspartate aminotransferase (EC 2.6.1.1)',\n",
       "        'Transaldolase (EC 2.2.1.2)',\n",
       "        '23S rRNA (cytosine(1962)-C(5))-methyltransferase (EC 2.1.1.191)',\n",
       "        'N-acetylglutamate kinase (EC 2.7.2.8)',\n",
       "        'Methyl-accepting chemotaxis protein',\n",
       "        'HAMP domain/GAF domain/HD domain protein',\n",
       "        '1,4-dihydroxy-2-naphthoate polyprenyltransferase (EC 2.5.1.74)',\n",
       "        'UDP-N-acetylenolpyruvoylglucosamine reductase (EC 1.3.1.98)',\n",
       "        'Universal stress protein UspA and related nucleotide-binding proteins',\n",
       "        'LSU ribosomal protein L7p/L12p (P1/P2)',\n",
       "        'LSU ribosomal protein L10p (P0)',\n",
       "        '2-keto-3-deoxy-D-arabino-heptulosonate-7-phosphate synthase I alpha (EC 2.5.1.54)',\n",
       "        \"3'-to-5' oligoribonuclease A, Bacillus type\",\n",
       "        'FIG000605: protein co-occurring with transport systems (COG1739)',\n",
       "        'Pyruvate-flavodoxin oxidoreductase',\n",
       "        'Multidrug resistance transporter, Bcr/CflA family',\n",
       "        'Sugar-phosphate guanylyltransferase / Sugar-phosephate isomerase',\n",
       "        'Oxidoreductase, short-chain dehydrogenase/reductase family',\n",
       "        'putative curli production assembly/transport component CsgG',\n",
       "        'alpha-xylosidase (EC 3.2.1.177)',\n",
       "        'Transcriptional regulator, AcrR family',\n",
       "        '1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)',\n",
       "        'LSU ribosomal protein L31p @ LSU ribosomal protein L31p, zinc-dependent',\n",
       "        'DNA primase DnaG', 'Putative RNA-binding protein YlmH',\n",
       "        'DNA polymerase III polC-type (EC 2.7.7.7)',\n",
       "        'Uridylate kinase (EC 2.7.4.22)',\n",
       "        'Acyl-phosphate:glycerol-3-phosphate O-acyltransferase PlsY (EC 2.3.1.n3)',\n",
       "        'ATP-dependent protease La (EC 3.4.21.53) Type I',\n",
       "        'Pyruvate formate-lyase activating enzyme (EC 1.97.1.4)',\n",
       "        'Pyruvate formate-lyase (EC 2.3.1.54)',\n",
       "        'Phosphomethylpyrimidine synthase ThiC (EC 4.1.99.17)',\n",
       "        'Triosephosphate isomerase (EC 5.3.1.1)',\n",
       "        'Glucosamine-1-phosphate guanylyltransferase',\n",
       "        'Manganese transport protein MntH',\n",
       "        'Pyruvate,phosphate dikinase (EC 2.7.9.1)',\n",
       "        'Maltose O-acetyltransferase (EC 2.3.1.79)',\n",
       "        'Homolog of fucose/glucose/galactose permeases',\n",
       "        'Oligopeptide transporter, OPT family',\n",
       "        'Inner membrane protein RclC associated with response to reactive chlorine species',\n",
       "        'ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17) @ ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17), cobalamin synthesis',\n",
       "        'Nicotinate phosphoribosyltransferase (EC 6.3.4.21)',\n",
       "        '6-phosphogluconolactonase (EC 3.1.1.31), eukaryotic type',\n",
       "        'Inner membrane protein YhiM', 'UPF0758 family protein',\n",
       "        'Predicted sialic acid transporter',\n",
       "        'Aldose 1-epimerase (EC 5.1.3.3)',\n",
       "        'Prolyl-tRNA synthetase (EC 6.1.1.15), archaeal/eukaryal type',\n",
       "        'Putative helicase',\n",
       "        '2-dehydro-3-deoxy-D-gluconate 5-dehydrogenase (EC 1.1.1.127) @ 2-deoxy-D-gluconate 3-dehydrogenase (EC 1.1.1.125)',\n",
       "        'Oxygen-insensitive NAD(P)H nitroreductase (EC 1.-.-.-) / Dihydropteridine reductase (EC 1.5.1.34)',\n",
       "        'Glycerol-3-phosphate transporter',\n",
       "        'DNA-binding transcriptional dual regulator SoxS',\n",
       "        'Thioredoxin reductase (EC 1.8.1.9)',\n",
       "        'Lysine/cadaverine antiporter membrane protein CadB',\n",
       "        'Low affinity tryptophan permease',\n",
       "        'PTS system, mannitol-specific IIC component / PTS system, mannitol-specific IIB component (EC 2.7.1.197) / PTS system, mannitol-specific IIA component (EC 2.7.1.197)',\n",
       "        'BarA-associated response regulator UvrY (= GacA = SirA)',\n",
       "        'Phosphoenolpyruvate-protein phosphotransferase of PTS system (EC 2.7.3.9)',\n",
       "        'Putative capsular polysaccharide transport protein YegH',\n",
       "        'Low-affinity inorganic phosphate transporter',\n",
       "        'Diguanylate cyclase (EC 2.7.7.65) => DgcC',\n",
       "        'Maltodextrin phosphorylase (EC 2.4.1.1)',\n",
       "        'Formate hydrogenlyase subunit 4',\n",
       "        'Putative efflux (PET) family inner membrane protein YccS',\n",
       "        'Galactose-1-phosphate uridylyltransferase (EC 2.7.7.10)',\n",
       "        'General secretion pathway protein M',\n",
       "        'Predicted sugar ABC transport system, ATP-binding protein YphE',\n",
       "        'DNA polymerase I (EC 2.7.7.7)', 'Uncharacterized protein YigF',\n",
       "        'Twin-arginine translocation protein TatA',\n",
       "        'Single-stranded DNA-binding protein',\n",
       "        'Inner membrane protein YejM, alkaline phosphatase superfamily',\n",
       "        'Inner membrane metabolite transport protein YgcS',\n",
       "        'ABC transporter, substrate-binding protein YnjB',\n",
       "        'Chorismate synthase (EC 4.2.3.5)',\n",
       "        'Outer membrane lipoprotein carrier protein LolA',\n",
       "        'Transcription antitermination protein NusG',\n",
       "        'Protein translocase subunit SecE',\n",
       "        'Uncharacterized iron-regulated membrane protein; Iron-uptake factor PiuB',\n",
       "        'Undecaprenyl diphosphate synthase (EC 2.5.1.31)',\n",
       "        'Nitroreductase', '6-phosphofructokinase (EC 2.7.1.11)',\n",
       "        'Phosphoribosylglycinamide formyltransferase (EC 2.1.2.2)',\n",
       "        'UDP-galactopyranose mutase (EC 5.4.99.9)',\n",
       "        'Alpha-glucosidase (EC 3.2.1.20)', 'TPR-domain-containing protein',\n",
       "        'Putative superfamily I DNA helicases',\n",
       "        'Uncharacterized ATPase, AAA superfamily',\n",
       "        'Alcohol dehydrogenase (EC 1.1.1.1)',\n",
       "        'O-acetylhomoserine sulfhydrylase (EC 2.5.1.49) @ O-succinylhomoserine sulfhydrylase (EC 2.5.1.48)',\n",
       "        'RNA polymerase ECF-type sigma factor',\n",
       "        'MutS-related protein, family 1',\n",
       "        'Tetracycline resistance, ribosomal protection type => Tet(Q)',\n",
       "        'UPF0056 inner membrane protein MarC', nan,\n",
       "        'Uncharacterized membrane protein PA1577',\n",
       "        'PAP2 superfamily protein', 'Sucrose permease',\n",
       "        '16S rRNA (guanine(527)-N(7))-methyltransferase (EC 2.1.1.170)',\n",
       "        'Peptidoglycan D,D-transpeptidase MrdA (EC 3.4.16.4)',\n",
       "        'Class A beta-lactamase (EC 3.5.2.6) => CfxA family, broad-spectrum',\n",
       "        'Vancomycin (or other glycopeptides) response regulator VanR => unclassified',\n",
       "        'Tetracycline resistance, tetracycline-inactivating enzyme => Tet(X)'],\n",
       "       dtype=object))"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter the data for rows related to \"Stress Response, Defense and Virulence\" and inspect unique values in the \"Role Name\" and \"Product\" columns\n",
    "stress_defense_virulence_data = Metagenomic[Metagenomic['Class'] == 'Stress Response, Defense and Virulence']\n",
    "\n",
    "# Analyze unique Role Names and Products in this filtered data\n",
    "unique_role_names = stress_defense_virulence_data['Role Name'].unique()\n",
    "unique_products = stress_defense_virulence_data['Product'].unique()\n",
    "\n",
    "unique_role_names, unique_products\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(       Subsystem Name                                            Product\n",
       " 398               NaN                 DNA gyrase subunit B (EC 5.99.1.3)\n",
       " 447               NaN       DNA topoisomerase IV subunit A (EC 5.99.1.3)\n",
       " 1006              NaN                 DNA gyrase subunit A (EC 5.99.1.3)\n",
       " 1012              NaN                         Aerotolerance protein BatE\n",
       " 1013              NaN                         Aerotolerance protein BatD\n",
       " ...               ...                                                ...\n",
       " 550339            NaN                 DNA gyrase subunit B (EC 5.99.1.3)\n",
       " 550343            NaN                 DNA gyrase subunit A (EC 5.99.1.3)\n",
       " 550362            NaN                 DNA gyrase subunit A (EC 5.99.1.3)\n",
       " 550363            NaN  Tetracycline resistance, ribosomal protection ...\n",
       " 550366            NaN  Tetracycline resistance, ribosomal protection ...\n",
       " \n",
       " [1875 rows x 2 columns],\n",
       " Product\n",
       " DNA gyrase subunit A (EC 5.99.1.3)                                                                   256\n",
       " DNA gyrase subunit B (EC 5.99.1.3)                                                                   249\n",
       " DNA topoisomerase IV subunit A (EC 5.99.1.3)                                                         172\n",
       " DNA topoisomerase IV subunit B (EC 5.99.1.3)                                                         158\n",
       " Choloylglycine hydrolase (EC 3.5.1.24)                                                               138\n",
       " MoxR-like ATPase in aerotolerance operon                                                              80\n",
       " Macrolide-specific ABC-type efflux carrier (TC 3.A.1.122.1)                                           73\n",
       " Manganese catalase (EC 1.11.1.6) => Spore coat protein CotJC                                          66\n",
       " Aerotolerance protein BatA                                                                            66\n",
       " Macrolide export ATP-binding/permease protein                                                         58\n",
       " Aerotolerance protein BatB                                                                            57\n",
       " Ribosome protection-type tetracycline resistance related proteins, group 2 / YacP-like NYN domain     56\n",
       " Aerotolerance protein BatC                                                                            55\n",
       " Tetracycline resistance, ribosomal protection type => Tet(Q)                                          54\n",
       " Aerotolerance protein BatD                                                                            49\n",
       " Aerotolerance protein BatE                                                                            43\n",
       " Tetracycline resistance element mobilization regulatory protein RteC                                  33\n",
       " Glutathione peroxidase (EC 1.11.1.9) @ Thioredoxin peroxidase (EC 1.11.1.15)                          27\n",
       " Tetracycline resistance, ribosomal protection type => Tet(W)                                          23\n",
       " Tetracycline resistance, ribosomal protection type => Tet(O)                                          23\n",
       " Choloylglycine hydrolase                                                                              17\n",
       " Catalase KatE (EC 1.11.1.6)                                                                           16\n",
       " Macrolide resistance, MFS efflux pump => Mef(A)                                                       15\n",
       " Sphingomyelinase (EC 3.1.4.12)                                                                        14\n",
       " Macrolide resistance, MFS efflux pump => Mef(En2)                                                     11\n",
       " Tetracycline resistance, tetracycline-inactivating enzyme => Tet(X)                                    9\n",
       " Aerotolerance protein BatD / Aerotolerance protein BatE                                                8\n",
       " Macrolide-specific efflux protein MacA                                                                 7\n",
       " ABC transporter membrane-spanning permease - macrolide efflux                                          5\n",
       " Aerotolerance protein BatB / Aerotolerance protein BatC                                                5\n",
       " Tetracycline resistance, ribosomal protection type => Tet(32)                                          5\n",
       " DNA gyrase inhibitor YacG                                                                              5\n",
       " Macrolide export ATP-binding/permease protein MacB                                                     4\n",
       " Tetracycline resistance, MFS efflux pump => Tet(40)                                                    4\n",
       " Manganese catalase (EC 1.11.1.6)                                                                       4\n",
       " Ribosome protection-type tetracycline resistance related proteins, group 2                             4\n",
       " Tetracycline resistance, ribosomal protection type => Tet(M)                                           1\n",
       " tetracycline resistance element mobilization regulatory protein rteC                                   1\n",
       " Macrolide glycosyltransferase (EC 2.4.1.-)                                                             1\n",
       " Catalase KatE-intracellular protease (EC 1.11.1.6)                                                     1\n",
       " Tetracycline resistance, ribosomal protection type => Tet(44)                                          1\n",
       " Catalase-peroxidase KatG (EC 1.11.1.21)                                                                1\n",
       " Name: count, dtype: int64)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extracting specific enzymes and associated bacteria names from \"Product\" and \"Subsystem Name\"\n",
    "# Filtering rows where we suspect relevance to ulcer-related bacterial functions\n",
    "\n",
    "# Keywords related to the enzymes and functions identified in the insights\n",
    "enzyme_keywords = [\"Tetracycline\", \"Macrolide\", \"Sphingomyelinase\", \"Choloylglycine hydrolase\", \"Catalase\", \n",
    "                   \"Glutathione Peroxidase\", \"DNA gyrase\", \"Topoisomerase IV\", \"Aerotolerance\"]\n",
    "\n",
    "# Filter rows containing any of the enzyme keywords in the Product column\n",
    "ulcer_related_data = Metagenomic[Metagenomic['Product'].str.contains('|'.join(enzyme_keywords), case=False, na=False)]\n",
    "\n",
    "# Count occurrences of unique enzymes and bacteria mentions to identify prevalence\n",
    "enzyme_counts = ulcer_related_data['Product'].value_counts()\n",
    "\n",
    "# Display the filtered data and enzyme occurrences for further insights\n",
    "ulcer_related_data[['Subsystem Name', 'Product']], enzyme_counts\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Product: Lists the enzymes and proteins identified in the dataset that are relevant to gut health and potentially linked to ulcer conditions. Examples include: \n",
    "1. DNA gyrase subunit A/B (EC 5.99.1.3): Enzymes critical for bacterial DNA replication and repair, highly prevalent in pathogens.\n",
    "2. Aerotolerance proteins (e.g., BatE, BatD): Enable bacteria to survive in variable oxygen environments.\n",
    "3. Tetracycline resistance proteins (e.g., Tet(Q)): Indicate antibiotic resistance, complicating treatment of bacterial infections.\n",
    "4. Sphingomyelinase (EC 3.1.4.12): Implicated in breaking down host membrane lipids, which could lead to tissue damage.\n",
    "5. Catalase (KatE): Frequently found in oxidative stress resistance (16 occurrences). This enzyme allows bacteria to detoxify reactive oxygen species, enabling survival in the gut.\n",
    "\n",
    "\n",
    "High Counts of Resistance Genes: Antibiotic resistance genes like Tetracycline resistance (Tet(Q)) and Macrolide resistance (Mef(A)) suggest the gut microbiota may harbor resistant pathogens, potentially complicating treatments for infections.\n",
    "\n",
    "\n",
    "Pathogenicity Indicators:\n",
    "Sphingomyelinase and Choloylglycine hydrolase are associated with host tissue damage and bile acid metabolism, processes that can irritate the gut lining and exacerbate ulcers.\n",
    "\n",
    "\n",
    "Aerotolerance proteins (BatA-E) are found frequently, pointing to adaptive bacteria capable of surviving in both oxygen-rich and anaerobic gut regions, some of which might be opportunistic pathogens.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To analyze potential links to gut ulcers in humans:\n",
    "\n",
    "Investigate rows mentioning gastrointestinal or pathogenic bacteria often associated with ulcers, such as Helicobacter pylori.\n",
    "\n",
    "Review pathways and roles tied to mechanisms of ulcer formation, like ammonia production, urease activity, and cytotoxin production.\n",
    "\n",
    "\n",
    "Highlight pathways or genes that could influence inflammation or ulcer development in the gut."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Superclass</th>\n",
       "      <th>Class</th>\n",
       "      <th>Subclass</th>\n",
       "      <th>Subsystem Name</th>\n",
       "      <th>Gene Count</th>\n",
       "      <th>Role Count</th>\n",
       "      <th>Active</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>...</th>\n",
       "      <th>E-value.1</th>\n",
       "      <th>Pathway ID.1</th>\n",
       "      <th>Pathway Name.1</th>\n",
       "      <th>Pathway Class.1</th>\n",
       "      <th>Annotation.3</th>\n",
       "      <th>Unique Genome Count.1</th>\n",
       "      <th>Unique Gene Count.1</th>\n",
       "      <th>Unique EC Count.1</th>\n",
       "      <th>EC Conservation.1</th>\n",
       "      <th>Gene Conservation.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32856</th>\n",
       "      <td>ENERGY</td>\n",
       "      <td>Energy and Precursor Metabolites Generation</td>\n",
       "      <td>Fermentation</td>\n",
       "      <td>Fermentations: Lactate</td>\n",
       "      <td>34.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>active</td>\n",
       "      <td>Bacteria SRX12677516</td>\n",
       "      <td>2.45914</td>\n",
       "      <td>2.45914.con.0001</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33124</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria SRX12677516</td>\n",
       "      <td>2.45914</td>\n",
       "      <td>2.45914.con.0002</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349731</th>\n",
       "      <td>ENERGY</td>\n",
       "      <td>Energy and Precursor Metabolites Generation</td>\n",
       "      <td>Central Metabolism</td>\n",
       "      <td>Citrate lyase</td>\n",
       "      <td>17.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>active</td>\n",
       "      <td>Bacteria SRX12677518</td>\n",
       "      <td>2.45923</td>\n",
       "      <td>2.45923.con.0001</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350155</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria SRX12677518</td>\n",
       "      <td>2.45923</td>\n",
       "      <td>2.45923.con.0002</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows Ã— 106 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Superclass                                        Class  \\\n",
       "32856      ENERGY  Energy and Precursor Metabolites Generation   \n",
       "33124         NaN                                          NaN   \n",
       "349731     ENERGY  Energy and Precursor Metabolites Generation   \n",
       "350155        NaN                                          NaN   \n",
       "\n",
       "                  Subclass          Subsystem Name  Gene Count  Role Count  \\\n",
       "32856         Fermentation  Fermentations: Lactate        34.0         7.0   \n",
       "33124                  NaN                     NaN         NaN         NaN   \n",
       "349731  Central Metabolism           Citrate lyase        17.0         6.0   \n",
       "350155                 NaN                     NaN         NaN         NaN   \n",
       "\n",
       "        Active                Genome  Genome ID         Accession  ...  \\\n",
       "32856   active  Bacteria SRX12677516    2.45914  2.45914.con.0001  ...   \n",
       "33124      NaN  Bacteria SRX12677516    2.45914  2.45914.con.0002  ...   \n",
       "349731  active  Bacteria SRX12677518    2.45923  2.45923.con.0001  ...   \n",
       "350155     NaN  Bacteria SRX12677518    2.45923  2.45923.con.0002  ...   \n",
       "\n",
       "       E-value.1  Pathway ID.1  Pathway Name.1 Pathway Class.1 Annotation.3  \\\n",
       "32856        NaN           NaN             NaN             NaN          NaN   \n",
       "33124        NaN           NaN             NaN             NaN          NaN   \n",
       "349731       NaN           NaN             NaN             NaN          NaN   \n",
       "350155       NaN           NaN             NaN             NaN          NaN   \n",
       "\n",
       "       Unique Genome Count.1  Unique Gene Count.1  Unique EC Count.1  \\\n",
       "32856                    NaN                  NaN                NaN   \n",
       "33124                    NaN                  NaN                NaN   \n",
       "349731                   NaN                  NaN                NaN   \n",
       "350155                   NaN                  NaN                NaN   \n",
       "\n",
       "        EC Conservation.1 Gene Conservation.1  \n",
       "32856                 NaN                 NaN  \n",
       "33124                 NaN                 NaN  \n",
       "349731                NaN                 NaN  \n",
       "350155                NaN                 NaN  \n",
       "\n",
       "[4 rows x 106 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter data for entries that may include bacteria associated with gut ulcers (e.g., Helicobacter pylori).\n",
    "# Checking if there's any mention of \"Helicobacter\" or similar keywords in the 'Genome Name' or 'Source Organism' columns.\n",
    "ulcer_related_data = Metagenomic[\n",
    "    Metagenomic['Genome Name'].str.contains(\"Helicobacter\", case=False, na=False) |\n",
    "    Metagenomic['Source Organism'].str.contains(\"Helicobacter\", case=False, na=False)\n",
    "]\n",
    "\n",
    "# Display the filtered data relevant to ulcer-causing bacteria.\n",
    "ulcer_related_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "ulcer_related_data.to_csv(\"Bacteria_associated_Ulcer.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze potential indirect mechanisms related to gut ulcer development, Focus on data entries that may play a role in:\n",
    "\n",
    "Toxin Production: Certain bacterial toxins can damage the gastric lining.\n",
    "\n",
    "Inflammatory Response: Bacterial products that activate immune responses may contribute to chronic inflammation, leading to ulceration.\n",
    "\n",
    "Urease Activity and pH Modulation: Some bacteria produce urease, which increases local pH and could disrupt the gut environment.\n",
    "\n",
    "Amino Acid Metabolism: Specific amino acid pathways, particularly involving lysine or arginine, can influence bacterial growth and immune evasion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subsystem Name</th>\n",
       "      <th>Role Name</th>\n",
       "      <th>Function</th>\n",
       "      <th>Pathway Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fatty Acid Biosynthesis cluster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1-deoxy-D-xylulose 5-phosphate reductoisomeras...</td>\n",
       "      <td>Fatty acid biosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Nudix proteins (nucleoside triphosphate hydrol...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dihydropteroate synthase (EC 2.5.1.15)</td>\n",
       "      <td>Glycosphingolipid biosynthesis - globo series</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Translation termination factors, bacterial</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glycosphingolipid biosynthesis - ganglio series</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Glycogen metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3-oxoacyl-[acyl-carrier-protein] synthase, KAS...</td>\n",
       "      <td>Phenylalanine, tyrosine and tryptophan biosynt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>De Novo Pyrimidine Synthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1-deoxy-D-xylulose 5-phosphate reductoisomeras...</td>\n",
       "      <td>Tetracycline biosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550347</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptozocin;[[(3R,4R,5S,6R)-3-(butanoylamino)...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550349</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDP-diacylglycerol--glycerol-3-phosphate 3-pho...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550353</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glycerophosphoryl diester phosphodiesterase (E...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550359</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glycerophosphoryl diester phosphodiesterase (E...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550372</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glycerophosphoryl diester phosphodiesterase (E...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5765 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Subsystem Name Role Name  \\\n",
       "4                         Fatty Acid Biosynthesis cluster       NaN   \n",
       "11      Nudix proteins (nucleoside triphosphate hydrol...       NaN   \n",
       "12             Translation termination factors, bacterial       NaN   \n",
       "14                                    Glycogen metabolism       NaN   \n",
       "16                           De Novo Pyrimidine Synthesis       NaN   \n",
       "...                                                   ...       ...   \n",
       "550347                                                NaN       NaN   \n",
       "550349                                                NaN       NaN   \n",
       "550353                                                NaN       NaN   \n",
       "550359                                                NaN       NaN   \n",
       "550372                                                NaN       NaN   \n",
       "\n",
       "                                                 Function  \\\n",
       "4       1-deoxy-D-xylulose 5-phosphate reductoisomeras...   \n",
       "11                 Dihydropteroate synthase (EC 2.5.1.15)   \n",
       "12                                                    NaN   \n",
       "14      3-oxoacyl-[acyl-carrier-protein] synthase, KAS...   \n",
       "16      1-deoxy-D-xylulose 5-phosphate reductoisomeras...   \n",
       "...                                                   ...   \n",
       "550347  Streptozocin;[[(3R,4R,5S,6R)-3-(butanoylamino)...   \n",
       "550349  CDP-diacylglycerol--glycerol-3-phosphate 3-pho...   \n",
       "550353  Glycerophosphoryl diester phosphodiesterase (E...   \n",
       "550359  Glycerophosphoryl diester phosphodiesterase (E...   \n",
       "550372  Glycerophosphoryl diester phosphodiesterase (E...   \n",
       "\n",
       "                                             Pathway Name  \n",
       "4                                 Fatty acid biosynthesis  \n",
       "11          Glycosphingolipid biosynthesis - globo series  \n",
       "12        Glycosphingolipid biosynthesis - ganglio series  \n",
       "14      Phenylalanine, tyrosine and tryptophan biosynt...  \n",
       "16                              Tetracycline biosynthesis  \n",
       "...                                                   ...  \n",
       "550347                                                NaN  \n",
       "550349                                                NaN  \n",
       "550353                                                NaN  \n",
       "550359                                                NaN  \n",
       "550372                                                NaN  \n",
       "\n",
       "[5765 rows x 4 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering the data for entries related to toxins, inflammatory response, urease activity, and amino acid metabolism.\n",
    "# Searching relevant keywords in 'Subsystem Name', 'Role Name', 'Function', and 'Pathway Name' to capture these areas.\n",
    "\n",
    "keywords = [\"toxin\", \"inflammation\", \"urease\", \"amino acid\", \"pH\", \"immune\"]\n",
    "\n",
    "# Filter the dataset for any rows containing these keywords in specific columns\n",
    "indirect_mechanism_data = Metagenomic[\n",
    "    Metagenomic['Subsystem Name'].str.contains('|'.join(keywords), case=False, na=False) |\n",
    "    Metagenomic['Role Name'].str.contains('|'.join(keywords), case=False, na=False) |\n",
    "    Metagenomic['Function'].str.contains('|'.join(keywords), case=False, na=False) |\n",
    "    Metagenomic['Pathway Name'].str.contains('|'.join(keywords), case=False, na=False)\n",
    "]\n",
    "\n",
    "# Display the filtered data relevant to indirect mechanisms in ulcer development.\n",
    "indirect_data_ulcer= indirect_mechanism_data[['Subsystem Name', 'Role Name', 'Function', 'Pathway Name']]\n",
    "\n",
    "indirect_data_ulcer\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identifying Bacteria and Gene Functions: search for bacteria or genes in the dataset that might be tied to ulcer pathogenesis and gut health, looking for known gut-associated bacterial groups.\n",
    "\n",
    "\n",
    "Pathway and Function Insights: provide an analysis of pathways and functions that contribute to ulcer pathogenesis based on the filtered data, especially pathways involved in inflammation, toxin production, and cellular adhesion.\n",
    "\n",
    "\n",
    "Enzymes and Associations: By reviewing specific enzymes related to these functions, we can gain insights into their role in gut health and their connection to bacterial species known to influence ulcer formation.\n",
    "\n",
    "\n",
    "Special Proteins and Functional Insights: I will assess any proteins that may play a role in gut health, especially those involved in bacterial survival in acidic environments or interactions with host tissues.\n",
    "\n",
    "\n",
    "Pathogenesis and Species Identification: Based on the roles, enzymes, and pathways identified, I will aim to link any relevant bacteria or bacterial species to ulcer pathogenesis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gut-related bacteria or gene functions:\n",
      "                                 Source Organism Role Name  \\\n",
      "349888  Bifidobacterium longum (strain NCC 2705)       NaN   \n",
      "475456                    Salmonella typhimurium       NaN   \n",
      "475571                    Salmonella typhimurium       NaN   \n",
      "475678                    Salmonella typhimurium       NaN   \n",
      "475869                    Salmonella typhimurium       NaN   \n",
      "476077                    Salmonella typhimurium       NaN   \n",
      "476547                          Escherichia coli       NaN   \n",
      "476584                          Escherichia coli       NaN   \n",
      "476764                          Escherichia coli       NaN   \n",
      "477376                          Escherichia coli       NaN   \n",
      "\n",
      "                                                 Function Pathway Name  \n",
      "349888  the acid resistance membrane protein (hded) fa...          NaN  \n",
      "475456  the type iii (virulence-related) secretory pat...          NaN  \n",
      "475571  the type iii (virulence-related) secretory pat...          NaN  \n",
      "475678  the type iii (virulence-related) secretory pat...          NaN  \n",
      "475869  the type iii (virulence-related) secretory pat...          NaN  \n",
      "476077  the type iii (virulence-related) secretory pat...          NaN  \n",
      "476547  the acid resistance membrane protein (hded) fa...          NaN  \n",
      "476584                     the pts glucitol (gut) family.          NaN  \n",
      "476764  the acid resistance membrane protein (hded) fa...          NaN  \n",
      "477376                     the pts glucitol (gut) family.          NaN  \n",
      "\n",
      "Pathways and Functional Insights:\n",
      "Empty DataFrame\n",
      "Columns: [Pathway Name, Pathway Class, Function]\n",
      "Index: []\n",
      "\n",
      "Specific Enzymes and Associations with Gut Health:\n",
      "                            Source Organism  \\\n",
      "82      Bacteroides fragilis (strain YCH46)   \n",
      "210            Bacteroides ovatus ATCC 8483   \n",
      "226                                     NaN   \n",
      "231                                     NaN   \n",
      "362                                     NaN   \n",
      "...                                     ...   \n",
      "536854                                  NaN   \n",
      "536855                                  NaN   \n",
      "536904                                  NaN   \n",
      "536917                                  NaN   \n",
      "536927                                  NaN   \n",
      "\n",
      "                                                Role Name  \\\n",
      "82                                                    NaN   \n",
      "210                                                   NaN   \n",
      "226                                                   NaN   \n",
      "231                                                   NaN   \n",
      "362                                                   NaN   \n",
      "...                                                   ...   \n",
      "536854      Lipopolysaccharide export system protein LptA   \n",
      "536855      Lipopolysaccharide export system protein LptA   \n",
      "536904                           FIG015287: Zinc protease   \n",
      "536917  Lipopolysaccharide export system permease prot...   \n",
      "536927      Lipopolysaccharide export system protein LptA   \n",
      "\n",
      "                                                  Product  \\\n",
      "82                    ADP-heptose:LPS heptosyltransferase   \n",
      "210              Tricorn protease homolog 1 (EC 3.4.21.-)   \n",
      "226     putative transmembrane transport protein, puta...   \n",
      "231     Lipopolysaccharide cholinephosphotransferase L...   \n",
      "362     Lipopolysaccharide core biosynthesis protein LpsA   \n",
      "...                                                   ...   \n",
      "536854      Lipopolysaccharide export system protein LptA   \n",
      "536855      Lipopolysaccharide export system protein LptA   \n",
      "536904                           FIG015287: Zinc protease   \n",
      "536917  Lipopolysaccharide export system permease prot...   \n",
      "536927      Lipopolysaccharide export system protein LptA   \n",
      "\n",
      "                                                 Function  \n",
      "82      the drug/metabolite transporter (dmt) superfam...  \n",
      "210              the major facilitator superfamily (mfs).  \n",
      "226                    DNA gyrase subunit B (EC 5.99.1.3)  \n",
      "231                       Translation elongation factor G  \n",
      "362                                                   NaN  \n",
      "...                                                   ...  \n",
      "536854                                                NaN  \n",
      "536855                                                NaN  \n",
      "536904                                                NaN  \n",
      "536917                                                NaN  \n",
      "536927                                                NaN  \n",
      "\n",
      "[3975 rows x 4 columns]\n",
      "\n",
      "Special Proteins Relevant to Gut Health:\n",
      "                            Source Organism  \\\n",
      "82      Bacteroides fragilis (strain YCH46)   \n",
      "362                                     NaN   \n",
      "691                                     NaN   \n",
      "2180                                    NaN   \n",
      "3842                                    NaN   \n",
      "...                                     ...   \n",
      "535240                                  NaN   \n",
      "535580                                  NaN   \n",
      "536048                                  NaN   \n",
      "536049                                  NaN   \n",
      "536071                                  NaN   \n",
      "\n",
      "                                          Role Name  \\\n",
      "82                                              NaN   \n",
      "362                                             NaN   \n",
      "691                                             NaN   \n",
      "2180                                            NaN   \n",
      "3842                                            NaN   \n",
      "...                                             ...   \n",
      "535240  Organic solvent tolerance protein precursor   \n",
      "535580                    LPS-assembly protein LptD   \n",
      "536048    Copper homeostasis protein CutF precursor   \n",
      "536049                    LPS-assembly protein LptD   \n",
      "536071                    LPS-assembly protein LptD   \n",
      "\n",
      "                                                  Product  \\\n",
      "82                    ADP-heptose:LPS heptosyltransferase   \n",
      "362     Lipopolysaccharide core biosynthesis protein LpsA   \n",
      "691     LPS-assembly protein LptD @ Organic solvent to...   \n",
      "2180      ATP-dependent Clp protease adaptor protein ClpS   \n",
      "3842    LPS-assembly protein LptD @ Organic solvent to...   \n",
      "...                                                   ...   \n",
      "535240  LPS-assembly protein LptD @ Organic solvent to...   \n",
      "535580  LPS-assembly protein LptD @ Organic solvent to...   \n",
      "536048  Copper homeostasis protein CutF precursor / Li...   \n",
      "536049  LPS-assembly protein LptD @ Organic solvent to...   \n",
      "536071  LPS-assembly protein LptD @ Organic solvent to...   \n",
      "\n",
      "                                                 Function  \n",
      "82      the drug/metabolite transporter (dmt) superfam...  \n",
      "362                                                   NaN  \n",
      "691                                                   NaN  \n",
      "2180                                                  NaN  \n",
      "3842                                                  NaN  \n",
      "...                                                   ...  \n",
      "535240                                                NaN  \n",
      "535580                                                NaN  \n",
      "536048                                                NaN  \n",
      "536049                                                NaN  \n",
      "536071                                                NaN  \n",
      "\n",
      "[328 rows x 4 columns]\n",
      "\n",
      "Species and Their Pathogenic Mechanisms:\n",
      "                                 Source Organism Role Name Pathway Name  \\\n",
      "349888  Bifidobacterium longum (strain NCC 2705)       NaN          NaN   \n",
      "475456                    Salmonella typhimurium       NaN          NaN   \n",
      "475571                    Salmonella typhimurium       NaN          NaN   \n",
      "475678                    Salmonella typhimurium       NaN          NaN   \n",
      "475869                    Salmonella typhimurium       NaN          NaN   \n",
      "476077                    Salmonella typhimurium       NaN          NaN   \n",
      "476547                          Escherichia coli       NaN          NaN   \n",
      "476584                          Escherichia coli       NaN          NaN   \n",
      "476764                          Escherichia coli       NaN          NaN   \n",
      "477376                          Escherichia coli       NaN          NaN   \n",
      "\n",
      "                                                 Function  \n",
      "349888  the acid resistance membrane protein (hded) fa...  \n",
      "475456  the type iii (virulence-related) secretory pat...  \n",
      "475571  the type iii (virulence-related) secretory pat...  \n",
      "475678  the type iii (virulence-related) secretory pat...  \n",
      "475869  the type iii (virulence-related) secretory pat...  \n",
      "476077  the type iii (virulence-related) secretory pat...  \n",
      "476547  the acid resistance membrane protein (hded) fa...  \n",
      "476584                     the pts glucitol (gut) family.  \n",
      "476764  the acid resistance membrane protein (hded) fa...  \n",
      "477376                     the pts glucitol (gut) family.  \n"
     ]
    }
   ],
   "source": [
    "# Identify gut-related bacteria or gene functions\n",
    "# Define keywords to filter gut-related terms\n",
    "gut_related_keywords = [\"pathogenesis\", \"virulence\", \"adhesion\", \"gut\", \"ulcer\", \"cytotoxin\", \"immune\", \"acid resistance\"]\n",
    "\n",
    "# Filter dataset for relevant keywords in columns related to functions and pathways\n",
    "gut_bacteria_data = Metagenomic[\n",
    "    Metagenomic['Role Name'].str.contains('|'.join(gut_related_keywords), case=False, na=False) |\n",
    "    Metagenomic['Function'].str.contains('|'.join(gut_related_keywords), case=False, na=False) |\n",
    "    Metagenomic['Pathway Name'].str.contains('|'.join(gut_related_keywords), case=False, na=False)\n",
    "]\n",
    "\n",
    "# Display results for gut-related bacteria or gene functions\n",
    "print(\"Gut-related bacteria or gene functions:\")\n",
    "print(gut_bacteria_data[['Source Organism', 'Role Name', 'Function', 'Pathway Name']])\n",
    "\n",
    "# Analyze pathways and functional insights\n",
    "# show pathways related to inflammation, acid resistance, or cellular adhesion\n",
    "pathway_keywords = [\"inflammation\", \"acid resistance\", \"adhesion\", \"toxin\"]\n",
    "pathway_insights = Metagenomic[\n",
    "    Metagenomic['Pathway Name'].str.contains('|'.join(pathway_keywords), case=False, na=False)\n",
    "]\n",
    "\n",
    "# Display pathway insights\n",
    "print(\"\\nPathways and Functional Insights:\")\n",
    "print(pathway_insights[['Pathway Name', 'Pathway Class', 'Function']])\n",
    "\n",
    "# Investigate specific enzymes and associations with bacterial groups\n",
    "# Show enzymes that might be relevant to gut health and ulcer pathogenesis\n",
    "enzyme_keywords = [\"urease\", \"protease\", \"cytotoxin\", \"lipopolysaccharide\", \"LPS\"]\n",
    "enzyme_data = Metagenomic[\n",
    "    Metagenomic['Role Name'].str.contains('|'.join(enzyme_keywords), case=False, na=False) |\n",
    "    Metagenomic['Product'].str.contains('|'.join(enzyme_keywords), case=False, na=False)\n",
    "]\n",
    "\n",
    "# Display specific enzymes related to gut health\n",
    "print(\"\\nSpecific Enzymes and Associations with Gut Health:\")\n",
    "print(enzyme_data[['Source Organism', 'Role Name', 'Product', 'Function']])\n",
    "\n",
    "# Explore special proteins for gut-related bacteria\n",
    "# Identify proteins that may influence gut health (e.g., adhesins, invasins)\n",
    "protein_keywords = [\"adhesion\", \"invasion\", \"toxins\", \"LPS\"]\n",
    "protein_data = Metagenomic[\n",
    "    Metagenomic['Role Name'].str.contains('|'.join(protein_keywords), case=False, na=False) |\n",
    "    Metagenomic['Product'].str.contains('|'.join(protein_keywords), case=False, na=False)\n",
    "]\n",
    "\n",
    "# Display special proteins relevant to gut health\n",
    "print(\"\\nSpecial Proteins Relevant to Gut Health:\")\n",
    "print(protein_data[['Source Organism', 'Role Name', 'Product', 'Function']])\n",
    "\n",
    "# Identify species and pathogenic mechanisms\n",
    "# Use the filtered data to link source organisms with gut ulcer-related mechanisms\n",
    "species_mechanisms = gut_bacteria_data[['Source Organism', 'Role Name', 'Pathway Name', 'Function']]\n",
    "\n",
    "# Display species and their related pathogenic mechanisms\n",
    "print(\"\\nSpecies and Their Pathogenic Mechanisms:\")\n",
    "print(species_mechanisms)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Key Organisms Identified:\n",
    "\n",
    "1. Bifidobacterium longum (strain NCC 2705): A beneficial gut bacterium known for promoting gut health.\n",
    "2. Salmonella typhimurium: A pathogenic bacterium associated with gastrointestinal infections.\n",
    "3. Escherichia coli: Includes both commensal and pathogenic strains, some of which are involved in gut-related diseases.\n",
    "4. Bacteroides fragilis: A common gut bacterium with roles in digestion and immune modulation but also capable of causing infections under certain conditions.\n",
    "\n",
    "\n",
    "\n",
    "Relevant Gene Functions and Pathways\n",
    "Acid Resistance: Bifidobacterium longum and Escherichia coli feature acid resistance mechanisms, such as membrane proteins (e.g., HdeD) that help bacteria survive in the acidic gut environment.\n",
    "\n",
    "Virulence Mechanisms: Salmonella typhimurium employs Type III secretion systems, a hallmark of virulence that allows bacteria to inject proteins into host cells, disrupting normal functions and causing inflammation.\n",
    "\n",
    "Metabolism and Transport: Some bacteria exhibit roles in glucitol metabolism (E. coli), highlighting their adaptation to diverse nutrient sources in the gut.\n",
    "\n",
    "\n",
    "\n",
    "Special Proteins and Products\n",
    "1. Lipopolysaccharide (LPS) biosynthesis and assembly proteins (e.g., LptA, LptD): Identified in species like Bacteroides fragilis. LPS is a critical component of the bacterial outer membrane, playing roles in immune evasion and inflammation.\n",
    "\n",
    "2. Zinc protease and ADP-heptose:LPS heptosyltransferase: Enzymes involved in bacterial survival and pathogenicity, found in Bacteroides fragilis.\n",
    "\n",
    "\n",
    "\n",
    "Potential Implications for Gut Ulcers\n",
    "\n",
    "Pathogens: The presence of virulent strains of Salmonella typhimurium and pathogenic E. coli suggests potential involvement in gut inflammation and damage, precursors to ulcers.\n",
    "\n",
    "Acid Resistance Mechanisms: Identified in E. coli and Bifidobacterium longum, these traits are critical for bacteria surviving in the stomach and upper GI tract, regions susceptible to ulceration.\n",
    "\n",
    "LPS and Immune Response: LPS from gram-negative bacteria can trigger strong immune responses, potentially leading to chronic inflammation and ulcer formation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identify Gut Related Bacteria through pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Functional Insights from LPS and Proteases:\n",
      "                       Source Organism  \\\n",
      "33152    Megamonas hypermegale ART12/1   \n",
      "98581   Streptococcus pneumoniae TIGR4   \n",
      "98592                              NaN   \n",
      "98669                    Streptococcus   \n",
      "98901                              NaN   \n",
      "...                                ...   \n",
      "536230                             NaN   \n",
      "536309                             NaN   \n",
      "536614                             NaN   \n",
      "536806                             NaN   \n",
      "536904                             NaN   \n",
      "\n",
      "                                                Role Name  \\\n",
      "33152                                                 NaN   \n",
      "98581   ATP-dependent Clp protease proteolytic subunit...   \n",
      "98592   Cell division-associated, ATP-dependent zinc m...   \n",
      "98669                           LPS-assembly protein LptD   \n",
      "98901                            FIG015287: Zinc protease   \n",
      "...                                                   ...   \n",
      "536230  ATP-dependent Clp protease ATP-binding subunit...   \n",
      "536309  Cell division-associated, ATP-dependent zinc m...   \n",
      "536614                           FIG015287: Zinc protease   \n",
      "536806  ATP-dependent Clp protease proteolytic subunit...   \n",
      "536904                           FIG015287: Zinc protease   \n",
      "\n",
      "                                                 Function  \\\n",
      "33152   the integral membrane caax protease-2 (caax pr...   \n",
      "98581                                                 NaN   \n",
      "98592                    Translation elongation factor Tu   \n",
      "98669   MULTISPECIES: macrolide efflux MFS transporter...   \n",
      "98901   3-oxoacyl-[acyl-carrier-protein] synthase, KAS...   \n",
      "...                                                   ...   \n",
      "536230                                                NaN   \n",
      "536309                                                NaN   \n",
      "536614                                                NaN   \n",
      "536806                                                NaN   \n",
      "536904                                                NaN   \n",
      "\n",
      "                           Pathway Name  \n",
      "33152                               NaN  \n",
      "98581           Sphingolipid metabolism  \n",
      "98592   Arginine and proline metabolism  \n",
      "98669           Caprolactam degradation  \n",
      "98901                               NaN  \n",
      "...                                 ...  \n",
      "536230                              NaN  \n",
      "536309                              NaN  \n",
      "536614                              NaN  \n",
      "536806                              NaN  \n",
      "536904                              NaN  \n",
      "\n",
      "[174 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Filter for entries that mention LPS or protease in the Role Name or Function columns\n",
    "lps_protease_data = Metagenomic[\n",
    "    Metagenomic['Role Name'].str.contains(\"LPS|protease\", case=False, na=False) |\n",
    "    Metagenomic['Function'].str.contains(\"LPS|protease\", case=False, na=False)\n",
    "]\n",
    "\n",
    "# Display LPS and protease-related entries\n",
    "print(\"Functional Insights from LPS and Proteases:\")\n",
    "print(lps_protease_data[['Source Organism', 'Role Name', 'Function', 'Pathway Name']])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pathways Associated with LPS and Proteases:\n",
      "                                    Pathway Name  \\\n",
      "33152                                        NaN   \n",
      "98581                    Sphingolipid metabolism   \n",
      "98592            Arginine and proline metabolism   \n",
      "98669                    Caprolactam degradation   \n",
      "98901                                        NaN   \n",
      "98929                                        NaN   \n",
      "98944                                        NaN   \n",
      "99011                                        NaN   \n",
      "475494                                       NaN   \n",
      "475621                                       NaN   \n",
      "529318                 One carbon pool by folate   \n",
      "529319                      Histidine metabolism   \n",
      "529341  Glycine, serine and threonine metabolism   \n",
      "\n",
      "                                    Pathway Class  \\\n",
      "33152                                         NaN   \n",
      "98581                            Lipid Metabolism   \n",
      "98592                       Amino Acid Metabolism   \n",
      "98669   Xenobiotics Biodegradation and Metabolism   \n",
      "98901                                         NaN   \n",
      "98929                                         NaN   \n",
      "98944                                         NaN   \n",
      "99011                                         NaN   \n",
      "475494                                        NaN   \n",
      "475621                                        NaN   \n",
      "529318       Metabolism of Cofactors and Vitamins   \n",
      "529319                      Amino Acid Metabolism   \n",
      "529341                      Amino Acid Metabolism   \n",
      "\n",
      "                                                 Function  \n",
      "33152   the integral membrane caax protease-2 (caax pr...  \n",
      "98581                                                 NaN  \n",
      "98592                    Translation elongation factor Tu  \n",
      "98669   MULTISPECIES: macrolide efflux MFS transporter...  \n",
      "98901   3-oxoacyl-[acyl-carrier-protein] synthase, KAS...  \n",
      "98929            D-alanine--D-alanine ligase (EC 6.3.2.4)  \n",
      "98944   Aminoglycoside 9-nucleotidyltransferase (EC 2....  \n",
      "99011                                                 NaN  \n",
      "475494  the outer membrane lipopolysaccharide export p...  \n",
      "475621           the rhomboid protease (rhomboid) family.  \n",
      "529318                  SSU ribosomal protein S10p (S20e)  \n",
      "529319  3-oxoacyl-[acyl-carrier-protein] synthase, KAS...  \n",
      "529341  Hydrogen peroxide-inducible genes activator =>...  \n"
     ]
    }
   ],
   "source": [
    "# Extract unique pathways related to LPS and proteases\n",
    "lps_protease_pathways = lps_protease_data[['Pathway Name', 'Pathway Class', 'Function']].drop_duplicates()\n",
    "\n",
    "# Display pathways associated with LPS and proteases\n",
    "print(\"Pathways Associated with LPS and Proteases:\")\n",
    "print(lps_protease_pathways)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hypothetical Gut Pathogens and Associated Enzymes/Proteins:\n",
      "Escherichia coli: LPS, protease\n",
      "Salmonella: LPS, protease\n",
      "Helicobacter pylori: urease, protease\n",
      "Campylobacter jejuni: cytotoxin, LPS\n"
     ]
    }
   ],
   "source": [
    "# Hypothetical mapping of enzymes/proteins to common gut pathogens\n",
    "hypothetical_pathogens = {\n",
    "    'Escherichia coli': ['LPS', 'protease'],\n",
    "    'Salmonella': ['LPS', 'protease'],\n",
    "    'Helicobacter pylori': ['urease', 'protease'],\n",
    "    'Campylobacter jejuni': ['cytotoxin', 'LPS']\n",
    "}\n",
    "\n",
    "# Display mapping for reference\n",
    "print(\"Hypothetical Gut Pathogens and Associated Enzymes/Proteins:\")\n",
    "for pathogen, functions in hypothetical_pathogens.items():\n",
    "    print(f\"{pathogen}: {', '.join(functions)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Summary of Pathogenic Mechanisms\n",
    "Pathogen\tKey Enzymes/Proteins\tMechanism of Action\tRole in Ulcers\n",
    "Escherichia coli\tLPS, protease\tTriggers inflammation and disrupts host proteins\tCan lead to mucosal damage and inflammation, precursors to ulcers\n",
    "Salmonella\tLPS, protease\tPromotes invasion, inflammation, and immune evasion\tChronic infections may impair the mucosal layer, predisposing to ulcer formation\n",
    "Helicobacter pylori\tUrease, protease\tNeutralizes acid, damages mucus and epithelial cells\tDirectly causes gastric and duodenal ulcers by disrupting protective barriers\n",
    "Campylobacter jejuni\tCytotoxin, LPS\tInduces cell death and inflammation\tSevere infections can cause intestinal lesions akin to ulcers"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
